File: app/src/main/java/com/pawels96/skyrimperkcalculator/models/Perk.java
package com.pawels96.skyrimperkcalculator.models;

import com.pawels96.skyrimperkcalculator.enums.IPerk;

import java.util.ArrayList;
import java.util.List;

public class Perk {

    public List<Perk> children = new ArrayList<>();
    public List<Perk> parents = new ArrayList<>();

    private IPerk perk;

    public Perk(IPerk iPerk) {
        perk = iPerk;
        maxState = perk.getPerkInfo().getSkillLevel().length;
    }

    public IPerk getPerk() {
        return perk;
    }

    private int state = 0;
    private int maxState;
    private boolean stateIncreasing = true;

    public boolean isSelected() {
        return state != 0;
    }

    public boolean isMultiState() {
        return maxState > 1;
    }

    public int getSkillLevel() {
        return perk.getPerkInfo().getSkillLevel()[state - 1];
    }

    public String getAllSkillLevels() {
        StringBuilder sb = new StringBuilder();

        int levels = perk.getPerkInfo().getSkillLevel().length;

        for (int i = 0; i < levels; i++) {
            sb.append(Integer.toString(perk.getPerkInfo().getSkillLevel()[i]));
            if (i < levels - 1)
                sb.append(", ");
        }
        return sb.toString();
    }

    public int getState() {
        return state;
    }

    public int getMaxState() {
        return maxState;
    }

    public String getStateAsString() {
        return " (" + Integer.toString(state) + "/" + Integer.toString(maxState) + ")";
    }

    private boolean hasParent() {
        return !parents.isEmpty();
    }

    public boolean hasChildren() {
        return !children.isEmpty();
    }

    public static void connectPerks(Perk start, Perk end) {
        start.children.add(end);
        end.parents.add(start);
    }

    private boolean isNoParentConnected() {
        for (Perk n : parents)
            if (n.isSelected())
                return false;
        return true;
    }

    public void setState(int st) {
        state = st;
        onStateChanged();
    }

    private void onStateChanged() {

        if (state == 0)
            stateIncreasing = true;
        else if (state == maxState)
            stateIncreasing = false;

        if (isSelected()) {
            if (!hasParent())
                return;
            if (isNoParentConnected())
                parents.get(0).setState(1);
        }
        else {

            if (!hasChildren())
                return;

            for (Perk n : children) {
                if (n.isNoParentConnected())
                    n.setState(0);
            }
        }
    }

    public void nextState() {

        if (stateIncreasing)
            state++;
        else
            state--;

        onStateChanged();
    }

    public static boolean areNodesSelected(Perk start, Perk end) {
        return start.isSelected() && end.isSelected();
    }

}

File: app/src/main/java/com/pawels96/skyrimperkcalculator/domain/Perk.kt
package com.pawels96.skyrimperkcalculator.domain

import java.io.Serializable

class Perk(val perk: IPerk) : Serializable {

    var children: MutableList<Perk> = mutableListOf()
    var parents: MutableList<Perk> = mutableListOf()

    var state = 0
        set(value) {
            field = value.coerceIn(0..maxState)
            onStateChanged()
        }

    val maxState: Int = perk.perkInfo.skillLevel.size

    var isStateIncreasing = true

    val isSelected: Boolean get() = state != 0

    val isMultiState: Boolean get() = maxState > 1

    val skillLevel: Int get() = perk.perkInfo.skillLevel[(state - 1).coerceAtLeast(0)]

    val allSkillLevels: String
        get() {
            val sb = StringBuilder()
            val levels: Int = perk.perkInfo.skillLevel.size
            for (i in 0 until levels) {
                sb.append(perk.perkInfo.skillLevel[i])
                if (i < levels - 1) sb.append(" / ")
            }
            return sb.toString()
        }

    val stateAsString: String get() = " ($state/$maxState)"

    val hasParent: Boolean get() = parents.isNotEmpty()

    val hasChildren: Boolean get() = children.isNotEmpty()

    private val isNoParentConnected: Boolean
        get() = parents.none { it.isSelected }

    private fun onStateChanged() {
        when (state) {
            0 -> isStateIncreasing = true
            maxState -> isStateIncreasing = false
        }

        if (isSelected) {
            if (hasParent && isNoParentConnected) {
                parents[0].state = 1
            }
        } else {
            for (child in children) {
                if (child.isNoParentConnected)  {
                    child.state = 0
                }
            }
        }
    }

    fun nextState() {
        if (isStateIncreasing) state++ else state--
    }

    companion object {
        fun connectPerks(parent: Perk, child: Perk) {
            parent.children.add(child)
            child.parents.add(parent)
        }

        @JvmStatic
        fun areNodesSelected(parent: Perk, child: Perk): Boolean {
            return parent.isSelected && child.isSelected
        }
    }

    override fun equals(other: Any?): Boolean {
        return this === other || (other is Perk && perk == other.perk && state == other.state)
    }
}